### PR TITLE
Refactor `create-slots`

### PR DESCRIPTION
**Pull request by arelra**
## Why?

`create-slots` does not create multiple slots but still returned an array.

`create-slots` logic is hard to follow and could do with a refactor with functions being split out and better naming.

Replaced with more verbose functions to aid readability.

## Does this change need to be reproduced in dotcom-rendering ?

- [x] No
- [ ] Yes (please indicate your plans for DCR Implementation)

### Tested

- [x] Locally
- [x] On CODE (optional)

### CODE REVIEW

File: static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts [hunk 1]
```diff
@@ -194,61 +193,103 @@ const createAdSlotElements = (
...
-	}
+): HTMLElement => {
+	const adSlotConfig = adSlotConfigs[type];
+	const slotName = options.name ?? adSlotConfig.name ?? type;
 
-	Object.keys(sizes).forEach((size) => {
-		sizeStrings[size] = sizes[size].join('|');
-	});
+	const sizeMappings = mergeSizeMappings(
+		adSlotConfigs[type].sizeMappings,
```
Comment (chrislomaxjones) [inline static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts:276]: Since you've already accessed `adSlotConfigs[type]` on L272 should it be this?
```suggestion
		adSlotConfig.sizeMappings,
```

File: static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts [hunk 2]
```diff
@@ -194,61 +193,103 @@ const createAdSlotElements = (
...
+			result[device] = sizes.join('|');
+			return result;
+		},
+		{},
+	);
+
+/**
+ * Prefix all object keys with data-${key}
+ */
+const createDataAttributes = (
```
Comment (chrislomaxjones) [inline static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts:257]: Similar to `createClasses` in this function you could also narrow the return type to 
```
Record<`data-${string}`, string>`
```
 but I think it's up to personal preference whether to use these types over just strings.

File: static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts [hunk 3]
```diff
@@ -194,61 +193,103 @@ const createAdSlotElements = (
...
 
-	adSlots.push(adSlot);
+	return adSlot;
+};
 
-	return adSlots;
+/**
+ * Split class names and prefix all with ad-slot--${className}
+ */
+const createClasses = (slotName: string, classes?: string): string[] =>
```
Comment (chrislomaxjones) [inline static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts:202]: Take or leave this suggestion but you could narrow the type here to something like (but it'll require an `as const` assertion in the function body):

```suggestion
const createClasses = (slotName: string, classes?: string): `ad-slot--${string}`[] =>
```

File: static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts [hunk 4]
```diff
@@ -194,61 +193,103 @@ const createAdSlotElements = (
...
+
+/**
+ * Given default size mappings and additional size mappings from
+ * the createAdSlot options parameter.
+ *
+ * 1. Check that the options size mappings use known device names
+ * 2. If so merge the size mappings
+ *
+ */
+const mergeSizeMappings = (
```
Comment (chrislomaxjones) [inline static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts:215]: This function won't deduplicate size mappings where merging? Does a name like merge imply that it might?

**Review by chrislomaxjones — COMMENTED** [review]
These changes have made this function so much clearer! I only had a couple of questions.

This also feels like a great candidate for moving out to commercial-core in the near-term as well.

File: static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts (hunk 3 above)
Comment (mxdvl) [reply static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts:202]: You don’t need the `as const` assertion, you can simply:

1. Pass the expected return type to the `map` like `map<`ad-slot--${string}`>((className: string) => `ad-slot--${className}`)`
2. Then, use `Array<`ad-slot--${string}`>` as it’s a “complex type”

File: static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts (hunk 4 above)
Comment (mxdvl) [reply static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts:215]: Maybe “join” is the more explicit word?

**Comment by mxdvl** [comment]
I think it’s worth mentioning when the original PR when multiple ad slots started to be returned: https://github.com/guardian/frontend/pull/19408. I think it had to do with either AMP (or [Blockthrough](https://github.com/guardian/frontend/pull/19275)… 🤷‍♂️)

File: static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts (hunk 3 above)
Comment (chrislomaxjones) [reply static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts:202]: That's much better @mxdvl

File: static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts (hunk 4 above)
Comment (arelra) [reply static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts:215]: Oh thats interesting, to me 'merge' doesn't imply dedupe but I can see that it could be misconstrued.
I've updated to 'concat' to reflect the actual operation - I hope that is clearer?

**Review by mxdvl — APPROVED** [review]

**Review by chrislomaxjones — APPROVED** [review]

File: static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts (hunk 4 above)
Comment (chrislomaxjones) [reply static/src/javascripts/projects/commercial/modules/dfp/create-slots.ts]: I think concat makes the most sense, since at the end of the day we're concatenating the sizes at each breakpoint.

**Comment by arelra** [comment]
> Excellent discovery on the fact that we only render one ad slot!
> 
> I think it’s worth mentioning when the original PR when multiple ad slots started to be returned: #19408. I think it had to do with either AMP or [Blockthrough](https://github.com/guardian/frontend/pull/19275)… 🤷‍♂️

Oh wow good digging! 🤔 

I can see that we no longer have any references to `Blockthrough` in the code, which it seems was the driver for an array of slots being introduced.

And afaics there are no consumers expecting an array so I think we should be okay.

I'll test it though :)

**Comment by arelra** [comment]
> These changes have made this function so much clearer! I only had a couple of questions.
> 
> This also feels like a great candidate for moving out to commercial-core in the near-term as well.

Thanks for the excellent comments and @mxdvl  too !

File: static/src/javascripts/projects/commercial/modules/liveblog-adverts.spec.js [hunk 5]
```diff
@@ -63,21 +75,23 @@ describe('Liveblog Dynamic Adverts', () => {
...
-			);
-			const candidates = document.querySelectorAll(
-				'.js-liveblog-body > *:nth-child(5n+1)',
+	it('should insert ad slots', async () => {
+		spaceFiller.fillSpace.mockImplementationOnce(
+			createFillSpaceMock([
+				document.querySelector('.x1'),
+				document.querySelector('.x12'),
+			]),
+		);
```
Comment (arelra) [inline static/src/javascripts/projects/commercial/modules/liveblog-adverts.spec.js:84]: fyi I added this `spaceFiller` mock which might be useful for other `spaceFiller` clients.

File: static/src/javascripts/projects/commercial/modules/liveblog-adverts.js [hunk 6]
```diff
@@ -121,15 +115,13 @@ export const init = () => {
 		return Promise.resolve();
 	}
 
-	fastdom
+	return fastdom
```
Comment (arelra) [inline static/src/javascripts/projects/commercial/modules/liveblog-adverts.js:118]: fyi this now returns the fastdom promise so we can chain the result in the test. afaics its not chained/awaited anywhere else.

File: static/src/javascripts/projects/commercial/modules/liveblog-adverts.js [hunk 7]
```diff
@@ -69,34 +69,28 @@ const getSlotName = (isMobile, slotCounter) => {
...
+		if (para && para.parentNode) {
+			const adSlot = createAdSlot('inline', {
+				name: getSlotName(isMobile, AD_COUNTER),
+				classes: 'liveblog-inline',
+			});
+			// insert the ad slot container into the DOM
+			para.parentNode.insertBefore(adSlot, para.nextSibling);
+			// load and display the advert via GAM
+			addSlot(adSlot, false);
+			AD_COUNTER += 1;
```
Comment (arelra) [inline static/src/javascripts/projects/commercial/modules/liveblog-adverts.js:86]: Tidied this up @chrislomaxjones

Comment (chrislomaxjones) [reply static/src/javascripts/projects/commercial/modules/liveblog-adverts.js:86]: Great,`slot` -> `para` makes this so much easier to understand.

File: static/src/javascripts/projects/commercial/modules/comment-adverts.ts [hunk 8]
```diff
@@ -8,53 +8,43 @@ import { commercialFeatures } from '../../common/modules/commercial/commercial-f
...
-				addSlot(adSlot, false);
-				void Promise.resolve(mediator.emit('page:commercial:comments'));
-			})
-	);
+	return fastdom
+		.mutate(() => {
+			commentMainColumn.classList.add('discussion__ad-wrapper');
+			if (
+				!config.get('page.isLiveBlog') &&
+				!config.get('page.isMinuteArticle')
```
Comment (mxdvl) [inline static/src/javascripts/projects/commercial/modules/comment-adverts.ts:37]:
```suggestion
				!config.get<boolean>('page.isLiveBlog') &&
				!config.get<boolean>('page.isMinuteArticle')
```

**Review by mxdvl — APPROVED** [review]

**Review by chrislomaxjones — APPROVED** [review]
Thanks for adding the liveblog test!

File: static/src/javascripts/projects/commercial/modules/liveblog-adverts.js (hunk 7 above)
Comment (mxdvl) [reply static/src/javascripts/projects/commercial/modules/liveblog-adverts.js:86]: Types would be great, here, wouldn’t they?

Can the paras be things like `<blockquote>`, `<embed>`, `<picture>` or other?

Comment (arelra) [reply static/src/javascripts/projects/commercial/modules/liveblog-adverts.js:86]: Types would be very helpful - yes!

`paras` can be any element in theory - it depends on the `rules` selector properties passed into `spaceFiller` e.g.

https://github.com/guardian/frontend/blob/498120e31f3a4463183df876b9fd2b4c70325f2d/static/src/javascripts/projects/commercial/modules/liveblog-adverts.js#L49-L60

`paras` is just the naming convention used in other `spaceFiller` clients so I followed that here.

Comment (chrislomaxjones) [reply static/src/javascripts/projects/commercial/modules/liveblog-adverts.js:86]: So `paras` are the candidates that win a round of Spacefinder, so `liveblocks` may be more accurate? `liveblocks` matches the DCR name for these elements, I'm not sure what they're called in frontend.

Comment (mxdvl) [reply static/src/javascripts/projects/commercial/modules/liveblog-adverts.js:86]: `elements` ?

Comment (arelra) [reply static/src/javascripts/projects/commercial/modules/liveblog-adverts.js:86]: `elements` makes sense to me as they could be element - but I assume in reality they are paragraph boundaries in most cases.

I think however this is venturing into another PR for a wider tidy up of space-finder nomenclature...?
